shell/platform/windows/key_event_handler.cc

```diff
@@ -81,25 +88,128 @@ int GetModsForKeyState() {
     mods |= kScrollLock;
   return mods;
 }
+
+// This uses event data instead of generating a serial number because
+// information can't be attached to the redispatched events, so it has to be
+// possible to compute an ID from the identifying data in the event when it is
+// received again in order to differentiate between events that are new, and
+// events that have been redispatched.
+//
+// Another alternative would be to compute a checksum from all the data in the
+// event (just compute it over the bytes in the struct, probably skipping
+// timestamps), but the fields used below are enough to differentiate them, and
+// since Windows does some processing on the events (coming up with virtual key
+// codes, setting timestamps, etc.), it's not clear that the redispatched
+// events would have the same checksums.
+uint64_t CalculateEventId(int scancode, int action, bool extended) {
+  // Calculate a key event ID based on the scan code of the key pressed,
+  // and the flags we care about.
+  return scancode | (((action == WM_KEYUP ? KEYEVENTF_KEYUP : 0x0) |
+                      (extended ? KEYEVENTF_EXTENDEDKEY : 0x0))
+                     << 16);
+}
+
 }  // namespace
 
-KeyEventHandler::KeyEventHandler(flutter::BinaryMessenger* messenger)
+KeyEventHandler::KeyEventHandler(flutter::BinaryMessenger* messenger,
+                                 KeyEventHandler::SendInputDelegate send_input)
     : channel_(
           std::make_unique<flutter::BasicMessageChannel<rapidjson::Document>>(
               messenger,
               kChannelName,
-              &flutter::JsonMessageCodec::GetInstance())) {}
+              &flutter::JsonMessageCodec::GetInstance())),
+      send_input_(send_input) {
+  assert(send_input != nullptr);
+}
 
 KeyEventHandler::~KeyEventHandler() = default;
 
 void KeyEventHandler::TextHook(FlutterWindowsView* view,
                                const std::u16string& code_point) {}
 
-void KeyEventHandler::KeyboardHook(FlutterWindowsView* view,
+KEYBDINPUT* KeyEventHandler::FindPendingEvent(uint64_t id) {
+  if (pending_events_.empty()) {
+    return nullptr;
+  }
+  for (auto iter = pending_events_.begin(); iter != pending_events_.end();
+       ++iter) {
+    if (iter->first == id) {
+      return &iter->second;
+    }
+  }
+  return nullptr;
+}
+
+void KeyEventHandler::RemovePendingEvent(uint64_t id) {
+  for (auto iter = pending_events_.begin(); iter != pending_events_.end();
+       ++iter) {
+    if (iter->first == id) {
+      pending_events_.erase(iter);
+      return;
+    }
+  }
+  std::cerr << "Tried to remove pending event with id " << id
+            << ", but the event was not found." << std::endl;
+}
+
+void KeyEventHandler::AddPendingEvent(uint64_t id,
+                                      int scancode,
+                                      int action,
+                                      bool extended) {
+  FML_DCHECK(pending_events_.size() < kMaxPendingEvents)
```

Comment:
Actually, I can't use this here because including FML is a layer violation.  I'm going to go back to the cerr output...